### ISSUE TITLE
bug(6to5@~2.13):  Array comprehension scope broken when the same variable name is used outside

### ISSUE THREAD

**Issue by douglasduteil** [6to5/6to5#538]
Hi

The scope of the array comprehension seems broken when **the same variable name is used outside**. 

Here is a dummy example : 

```js
var foo = [];
[for (foo of bar) if (foo) foo]
```

Becomes

```js
"use strict";

var foo = [];
(function () {
  var _ref = [];

  for (var _iterator = bar[Symbol.iterator](), _step; !(_step = _iterator.next()).done;) {
    var _foo = _step.value;
    if (foo) {
      _ref.push(foo);
    }
  }

  return _ref;
})();
```

Instead of :

```js
"use strict";

var foo = [];
(function () {
  var _ref = [];

  for (var _iterator = bar[Symbol.iterator](), _step; !(_step = _iterator.next()).done;) {
    var _foo = _step.value;
    if (_foo) {
      _ref.push(_foo);
    }
  }

  return _ref;
})();
```

Notice the inner `if` statement and the `push` with `foo` instead of `_foo`

Here is a [working demo](https://6to5.org/repl/#?experimental=true&playground=true&evaluate=true&loose=false&code=%0A%2F%2F%20Messing%20up%20since%206to5%40~2.13%0A%0Alet%20c%20%3D%20%5B%5D%3B%0A%0Alet%20customers%20%3D%20%5B%20%7B%20city%3A%20'Paris'%2C%20name%3A%20'foo'%2C%20age%3A%2042%7D%2C%20%20%7B%20city%3A%20'Seattle'%2C%20name%3A%20'oof'%2C%20age%3A%2024%7D%5D%0Alet%20results%20%3D%20%5Bfor%20(c%20of%20customers)%20if%20(c.city%20%3D%3D%20%22Seattle%22)%20%7B%20name%3A%20c.name%2C%20age%3A%20c.age%20%7D%5D%0A%0Aconsole.log(results)%3B%0A)

**Comment by douglasduteil** [comment]
Might be linked to #537

**Comment by sebmck** [comment]
@douglasduteil Partially related, they're both due to bugs in the scope tracking.

**Comment by monsanto** [comment]
Just ran into this. It's a regression in version 2.13.0. It affects `for` loops in general.  Here is a simpler example:

```
let x = 0
for (;;) {
    let x = 1
    x
}
```

-->

```
var x = 0;
for (;;) {
  var _x = 1;
  x;
}
```

**Comment by sebmck** [comment]
Looks like it's due to my partial rewrite of the let scoping transformer due to #510. As per that issue:

> If some of you can write up some tests for these issues it'd be extremely appreciated. Basically something like [this](https://github.com/6to5/6to5/blob/master/test/fixtures/transformation/es6-let-scoping/exec-duplicate-in-upper-scope/exec.js) or [this](https://github.com/6to5/6to5/blob/master/test/fixtures/transformation/es6-let-scoping/exec-for-loop-head/exec.js).

**Comment by sebmck** [comment]
26cb5d5a65df70628b6e2bb7f392b5c6c1a7d736 only solves https://github.com/6to5/6to5/issues/538#issuecomment-70607198 and not the original issue.

**Comment by sebmck** [comment]
I fixed this specific issue, not sure how, I'm going to have to investigate to find out the actual bug/cause because I'm not going to feel comfortable releasing this hacky fix.

**Comment by sebmck** [comment]
Fixed as of 2.13.4, thanks!

**Comment by douglasduteil** [comment]
:+1:
